### PR TITLE
[GC] Removed unreferenced event Q for summarizer and log unreferenced events only if node was not revived first

### PR DESCRIPTION
**Pull request by agarwal-navin**
## Reviewer guidance

This is targetting a test branch (test/gc-2.3). Since this has legacy alpha API breaking changes which are not allowed to be merged until v2.3, this is a workaround to get the changes reviewed and merged in a test branch. Later when breaking changes are allowed, these will get merged to main.

## Description
In summarizer clients, unreferenced events (inactive, tombstone ready and sweep ready) are added to a queue and are logged after the next GC completes. This was done to reduce false positives in 2 scenarios:
- A node is used after it is revived. Since the revive notifcation to GC was delayed until the object reviving was loaded, GC could not build correct timeline.
- GC may get multiple reference notifications on object loads. GC was notified of references whenever an object with handles was loaded.

With the reference op work, both of these problems don't exist. So, we can remove the unreferenced event queue for summarizer clients. Also, we can check that an object is not revived before we log an event for it which will reduce false positives for non-summarizer clients too.

This PR makes 3 main changes:
1. It removes the unreferenced event queue for summarizers and logs unreferenced events right away.
2. It checks that GC did not see a reference for an unreferenced object before it was used. If it did, it does not log because the usage isn't unexpected.
3. Added packagePath and fromPackagePath properties to addedGCOutboundRoute. Since this information is already present with the runtime when it calls GC, it doesn't need to ask for it.

[AB#8938](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8938)

### CODE REVIEW

**Comment by agarwal-navin** [comment]
@microsoft/fluid-cr-api Note that this is against a test branch and not main.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -3463,11 +3450,41 @@ export class ContainerRuntime
...
+			timestampMs,
+		});
+	}
+
+	/**
+	 * Called to retrieve the package path of the node with the given path. The node should be root or belong to a
+	 * data store or an attachment blob.
+	 */
+	public getGCNodePackagePath(nodePath: string): readonly string[] | undefined {
+		// GC uses "/" when adding "root" references, e.g. for Aliasing or as part of Tombstone Auto-Recovery.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3473]: Why did you move this?

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 2]
```diff
@@ -97,8 +79,8 @@ export class GCTelemetryTracker {
 	// Keeps track of unreferenced events that are logged for a node. This is used to limit the log generation to one
 	// per event per node.
 	private readonly loggedUnreferencedEvents: Set<string> = new Set();
-	// Queue for unreferenced events that should be logged the next time GC runs.
-	private pendingEventsQueue: IUnreferencedEventProps[] = [];
+
+	private readonly referencesSinceLastRun: Set<string> = new Set();
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:83]: comment please

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 3]
```diff
@@ -97,8 +79,8 @@ export class GCTelemetryTracker {
 	// Keeps track of unreferenced events that are logged for a node. This is used to limit the log generation to one
 	// per event per node.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:80]: per session? Because it's not cleared in `gcRunCompleted`. Would be nice to confirm/clarify

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 4]
```diff
@@ -922,12 +915,12 @@ export class GarbageCollector implements IGarbageCollector {
...
-					"/",
-					tombstonedNodePath,
-					messageTimestampMs,
-					true /* autorecovery */,
-				);
+				this.addedOutboundReference({
+					fromNodePath: "/",
+					toNodePath: tombstonedNodePath,
+					timestampMs: messageTimestampMs,
+					autorecovery: true /* autorecovery */,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:922]: nit!
```suggestion
					autorecovery: true,
```

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 5]
```diff
@@ -1030,7 +1023,6 @@ export class GarbageCollector implements IGarbageCollector {
 			lastSummaryTime: this.getLastSummaryTimestampMs(),
 			headers: headerData,
 			requestUrl: request?.url,
-			requestHeaders: JSON.stringify(request?.headers),
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:1033]: I feel like this was here for a reason. It shows what was actually provided by caller, v. whatever headerData was modified for implementing logic.

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 6]
```diff
@@ -186,27 +181,49 @@ export class GCTelemetryTracker {
...
-					: -1,
-			timeout,
-			isTombstoned,
+
+		const loggingProps: ITelemetryPropertiesExt = {
 			...tagCodeArtifacts({ id: untaggedId, fromId: untaggedFromId }),
-			...otherNodeUsageProps,
-			...this.createContainerMetadata,
+			...tagCodeArtifacts({ pkg: packagePath?.join("/") }),
+			...tagCodeArtifacts({ fromPkg: fromPackagePath?.join("/") }),
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:188]: How about this?
```suggestion
			...tagCodeArtifacts({
				id: untaggedId,
				fromId: untaggedFromId,
				pkg: packagePath?.join("/"),
				fromPkg: fromPackagePath?.join("/"),
			}),
```

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 7]
```diff
@@ -218,86 +235,50 @@ export class GCTelemetryTracker {
...
+		loggingProps: ITelemetryPropertiesExt,
+		allowTombstone?: boolean,
 	) {
 		// This will log the following events:
 		// GC_Tombstone_DataStore_Requested, GC_Tombstone_DataStore_Changed, GC_Tombstone_DataStore_Revived
 		// GC_Tombstone_SubDataStore_Requested, GC_Tombstone_SubDataStore_Changed, GC_Tombstone_SubDataStore_Revived
 		// GC_Tombstone_Blob_Requested, GC_Tombstone_Blob_Changed, GC_Tombstone_Blob_Revived
-		const { id, fromId, headers, gcConfigs, additionalProps, ...detailedProps } =
-			unrefEventProps;
 		const eventUsageName = usageType === "Loaded" ? "Requested" : usageType;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:268]: The way you've changed this, it is going to pull a bunch of stuff that used to be in `details` up to the top level.  I'd like to leave it as-is, with key props at top level and the rest in `details`.  You can do the same destructuring trick here on `loggingProps`, to separate out the top-level stuff from the "extra" stuff to go into `details`.

I'm not sure why you removed the `IUnreferencedEventProps` type, since it gave some type safety on the props destructured here, I would have left this part totally as-is.  Maybe I'm missing some other key part of this change.

```suggestion
		const { id, fromId, headers, gcConfigs, additionalProps, ...detailedProps } =
			loggingProps;
		const eventUsageName = usageType === "Loaded" ? "Requested" : usageType;
```

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 8]
```diff
@@ -109,11 +91,16 @@ export class GCTelemetryTracker {
...
-			nodePath: string,
-		) => Promise<readonly string[] | undefined>,
 	) {}
 
+	/**
+	 * Called after every GC run. Clear all references since the last GC run since the unreferenced
+	 * state of all nodes would have updated and the references are now stale.
+	 */
+	public gcRunCompleted() {
+		this.referencesSinceLastRun.clear();
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:101]: How about just passing in `newReferencesSinceLastRun` from GC into `nodeUsed`, instead of tracking this thing?  Seems likely to be redundant.

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 9]
```diff
@@ -186,27 +181,49 @@ export class GCTelemetryTracker {
...
+			this.logTombstoneUsageTelemetry(
+				nodeType,
+				usageType,
+				loggingProps,
+				headers?.allowTombstone,
+			);
+		}
+
+		// If the node has been referenced before it was used, this usage is not unexpected. Check the trackedId
+		// as well for cases where a DDS is used after it's data store is referenced.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:215]: grammar
```suggestion
		// as well for cases where a DDS is used after its data store is referenced.
```

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 10]
```diff
@@ -186,27 +181,49 @@ export class GCTelemetryTracker {
...
+		if (
+			this.referencesSinceLastRun.has(untaggedId) ||
+			this.referencesSinceLastRun.has(trackedId)
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:218]: Feels like a bummer to have to account for this here. Feels like the DDS story is really shaky. I guess that's probably true because it's halfway supported.

If a DDS is used after its data store is referenced, the DDS is known to be referenced because of the implicit references from DataStore to DDS.

Let's think of the general case -- What if A -> B (both DataStores) but A is unreferenced, then outbound route is added to A, now B is referenced too. Do we handle that general case?  Can we?

If not, then I guess this is fine.  If we can handle that general case, then this isn't needed.

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 11]
```diff
@@ -218,86 +235,50 @@ export class GCTelemetryTracker {
...
-					eventName: `${state}Object_${usageType}`,
-					...tagCodeArtifacts({ pkg: packagePath?.join("/") }),
-					stack: generateStack(),
-					id,
-					fromId,
-					headers: { ...headers },
-					details: { ...detailedProps, ...additionalProps },
-					gcConfigs,
-				};
+		if (this.isSummarizerClient || usageType === "Loaded") {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:245]: Let's keep that comment about why interactive client log loaded events only.

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 12]
```diff
@@ -362,62 +343,6 @@ export class GCTelemetryTracker {
 			}
 		}
 	}
-
-	/**
-	 * Log events that are pending in pendingEventsQueue. This is called after GC runs in the summarizer client
-	 * so that the state of an unreferenced node is updated.
-	 */
-	public async logPendingEvents(logger: ITelemetryLoggerExt) {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:370]: I didn't compare this against the code above - was there anything else different in here? Or was it just basically duplicated?

File: packages/runtime/container-runtime/src/gc/gcConfigs.ts [hunk 13]
```diff
@@ -131,7 +131,7 @@ export function generateGCConfigs(
 		: "NO";
 
 	// If we aren't running sweep, also disable AutoRecovery which also emits the GC op.
-	// This gives a simple control surface for compability concerns around introducing the new op.
+	// This gives a simple control surface for compatibility concerns around introducing the new op.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcConfigs.ts:134]: (This will probably conflict with that configs PR, maybe move it there...)

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 14]
```diff
@@ -737,9 +737,23 @@ export abstract class FluidDataStoreContext
...
+		toPath: string,
+		messageTimestampMs?: number,
+		fromPkg?: readonly string[],
+		toPkg?: readonly string[],
+	) {
+		this.parentContext.addedGCOutboundRoute(
+			fromPath,
+			toPath,
+			messageTimestampMs,
+			fromPkg ?? this.pkg,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:754]: Should we just always use `this.pkg`? When would the provided value be different?

File: packages/runtime/container-runtime/src/channelCollection.ts [hunk 15]
```diff
@@ -1406,15 +1428,17 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
...
-	): Promise<readonly string[] | undefined> {
+	public getDataStorePackagePath(nodePath: string): readonly string[] | undefined {
 		// If the node belongs to a data store, return its package path. For DDSes, we return the package path of the
 		// data store that contains it.
-		// TODO why are we non null asserting here?
-		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-		const context = this.contexts.get(nodePath.split("/")[1]!);
-		return (await context?.getInitialSnapshotDetails())?.pkg;
+		const dataStoreId = nodePath.split("/")[1];
+		if (dataStoreId !== undefined) {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/channelCollection.ts:1435]: Maybe mention in the doc comment that if you pass the root path it'll return undefined?  And I guess mention the other case where it returns undefined which is if it's for a context that's not loaded yet.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 14 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:754]: Actually now I'm wondering the flip side - why is this a legitimate default?  Brain is slowing down so maybe it'll be obvious next time I look at this...

Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:754]: The only usage I can see for DataStoreContext is when processing a DDS attach message, and neither pkg is passed.  This is fine, ContainerRuntime will fill them in.  I would just not even include them in the signature of this implementation - it will still satisfy the interface... although it means that if someone _does_ pass them they'll be ignored.  Hmm I guess it's ok to pass them through, but it feels cleaner somehow to not and just say ContainerRuntime will compute them.

Adding this fallback here feels more confusing - what scenario is it actually helpful?  If someone passes it, use it. If not, fall back to the common computation in ContainerRuntime.  Less code/cases is better.

Thoughts?

File: packages/test/test-end-to-end-tests/src/test/gc/gcInactiveNodes.spec.ts [hunk 16]
```diff
@@ -517,11 +514,9 @@ describeCompat("GC inactive nodes tests", "NoCompat", (getTestObjectProvider, ap
 			itExpects(
 				"throwOnInactiveLoad: true; DDS handle.get -- Doesn't throw, and DOESN'T log",
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcInactiveNodes.spec.ts:515]: Update test name.  Nice to see this fixed.

File: packages/test/test-end-to-end-tests/src/test/gc/gcInactiveNodes.spec.ts [hunk 17]
```diff
@@ -569,30 +564,28 @@ describeCompat("GC inactive nodes tests", "NoCompat", (getTestObjectProvider, ap
 					);
 
 					// Bug: It SHOULD actually log
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcInactiveNodes.spec.ts:566]:
```suggestion
```

File: packages/runtime/container-runtime/src/test/gc/gcTelemetry.spec.ts [hunk 18]
```diff
@@ -514,28 +501,26 @@ describe("GC Telemetry Tracker", () => {
...
 					clock.tick(timeout - 1);
 					mockNodeChanges(nodes);
-					await simulateGCToTriggerEvents(isSummarizerClient);
 
 					validateNoEvents();
 
 					// Expire the timeout and validate that only revived event is generated for node 2.
 					clock.tick(1);
+					mockNodeRevival(nodes[1], nodes[2]);
 					mockNodeChanges([nodes[2]]);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/gcTelemetry.spec.ts:523]: Switching order here so that the revival happens before the change, since we're logging directly instead of queuing?

**Review by markfields — APPROVED** [review]
Reviewed all the code pretty closely, more skimmed the tests but I liked what I saw.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 14 above)
Comment (rajatch-ff) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:754]: Yeah I have that doubt too. In `addedGCOutboundRoute` to begin with. We accept both fromPath and fromPkg, and eventually do this
> const packagePath = toPkg ?? this.getGCNodePackagePath(toPath);

Curious what's the advantage of also allowing caller to send the frompkg and topkg, when we already have the frompath and to path!

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts [hunk 19]
```diff
@@ -31,50 +29,34 @@ import { UnreferencedStateTracker } from "./gcUnreferencedStateTracker.js";
...
 	requestUrl?: string;
+	/** The timestamp of the last summary */
+	lastSummaryTime?: number;
 	/** Original request headers if this usage came from a request or handle.get */
-	requestHeaders?: string;
+	headers?: RuntimeHeaderData;
+	/** Additional properties to log in telemetry */
+	additionalProps?: ITelemetryPropertiesExt;
+	/** The package path of the node that adds a reference. Not available if the node hasn't been loaded yet */
+	fromPackagePath?: readonly string[] | undefined;
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/src/gc/gcTelemetry.ts:59]: Consider moving it up, close to `packagePath`?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3473]: Just to make it next to where it's used.

File: packages/runtime/container-runtime/api-report/container-runtime.legacy.alpha.api.md [hunk 20]
```diff
@@ -37,7 +37,7 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents
...
     // (undocumented)
     protected addContainerStateToSummary(summaryTree: ISummaryTreeWithStats, fullTree: boolean, trackState: boolean, telemetryContext?: ITelemetryContext): void;
-    addedGCOutboundRoute(fromPath: string, toPath: string, messageTimestampMs?: number): void;
+    addedGCOutboundRoute(fromPath: string, toPath: string, messageTimestampMs?: number, fromPkg?: readonly string[], toPkg?: readonly string[]): void;
```
[the remark this answers is not in the document]
Comment (agarwal-navin) [reply packages/runtime/container-runtime/api-report/container-runtime.legacy.alpha.api.md:40]: The only place these may not be provided is when it's called by the data store context directly when it processes an attach message for a DDS. In that case, it should still have `fromPkg` because that's the data store itself. `toPkg` can be undefined in that case. 

However, this function is part of `IFluidParentContext` which is implemented also implemented by `FluidDataStoreContext` where `fromPkg` also needs to be optional.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 14 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:754]: I guess to keep things simple, we could leave `addedGCOutboundRoute` unchanged and just have container runtime get the package path when calling GarbageCollector. The goal was to not have the GarbageCollector try to get it.

The reason I added package paths everywhere is for optimization, i.e., if the call is coming directly from a data store or when processing data store attach message, the package path is readily available so there is no need to get it again. But getting the package is cheap enough so we should not add this complication.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:1033]: We never looked at it and I doubt we will. The only case where headerData and headers can be different is when requesting for sub-data store objects. Do we need to differentiate that case?

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/gcTelemetry.ts:83]: The reason I didn't add comment because it's pretty clear from the name. Not sure what additional information I should add.

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/gcTelemetry.ts:101]: That is what I started with. But `newReferencesSinceLastRun` is not a list of references. It is a list of `nodeId -> references`. So, we will have to get the references from it every time `nodeUsed` is called which is for every op making it very expensive.

File: packages/runtime/container-runtime/src/gc/gcTelemetry.ts (hunk 10 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/gcTelemetry.ts:218]: Good point. This does change the behavior. Before this change, after GC runs, B will be referenced and we will not log if it was used. With this change, we will log for B.

I don't think there is a straightforward way to handle this case. We could, but it will be more complicated than it already is. So, I think this change may not be worth it.

**Comment by agarwal-navin** [comment]
As discussed with @markfields offline and per this comment https://github.com/microsoft/FluidFramework/pull/21961#discussion_r1689055623, closing this PR for now. This can regress behaviors and can be more costly than it's worth. We can revisit this later and remove the whole thing once GC is stable. We will only need tombstone and deleted logs then.
